serde_v8/src/magic/transl8.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+//! Transerialization extends the set of serde-compatible types (for given de/serializers).
+//! By "hackishly" transmuting references across serde boundaries as u64s.
+//! Type-safety is enforced using special struct names for each "magic type".
+//! Memory-safety relies on transerialized values being "pinned" during de/serialization.
+
+pub(crate) const MAGIC_FIELD: &str = "$__v8_magic_field";
+
+pub(crate) trait MagicType {
+  const NAME: &'static str;
+  const MAGIC_NAME: &'static str;
+}
+
+pub(crate) trait ToV8 {
+  fn to_v8<'a>(
+    &self,
+    scope: &mut v8::HandleScope<'a>,
+  ) -> Result<v8::Local<'a, v8::Value>, crate::Error>;
+}
+
+pub(crate) trait FromV8: Sized {
+  fn from_v8(
+    scope: &mut v8::HandleScope,
+    value: v8::Local<v8::Value>,
+  ) -> Result<Self, crate::Error>;
+}
+
+pub(crate) fn magic_serialize<T, S>(
+  serializer: S,
+  x: &T,
+) -> Result<S::Ok, S::Error>
+where
+  S: serde::Serializer,
+  T: MagicType,
+{
+  use serde::ser::SerializeStruct;
+
+  let mut s = serializer.serialize_struct(T::MAGIC_NAME, 1)?;
+  let ptr = opaque_send(x);
+  s.serialize_field(MAGIC_FIELD, &ptr)?;
+  s.end()
+}
+
+pub(crate) fn magic_deserialize<'de, T, D>(
+  deserializer: D,
+) -> Result<T, D::Error>
+where
+  D: serde::Deserializer<'de>,
+  T: MagicType,
+{
+  struct ValueVisitor<T> {
+    p1: std::marker::PhantomData<T>,
+  }
+
+  impl<'de, T: MagicType> serde::de::Visitor<'de> for ValueVisitor<T> {
+    type Value = T;
+
+    fn expecting(
+      &self,
+      formatter: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+      formatter.write_str("a ")?;
+      formatter.write_str(T::NAME)
+    }
+
+    fn visit_u64<E>(self, ptr: u64) -> Result<Self::Value, E>
+    where
+      E: serde::de::Error,
+    {
+      // SAFETY: opaque ptr originates from visit_magic, which forgets ownership so we can take it
+      Ok(unsafe { opaque_take(ptr) })
+    }
```

Comment:
I might be completely missing the point here - but is this assuming that pointers can be passed to JavaScript as 64 bit integers?

Reply:
No it's not, we're simply transmuting pointers/references across the "serde boundary" as opaque u64s so our specialized de/serializer can pick it up

This refactor doesn't change any of serde_v8's observable qualities, it streamlines magic types by formalizing transerialization making them significantly easier to reason about, maintain, evolve or extend.